pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,285 @@
+     * price changes and BPT price changes. This calculation transforms one into the other.
+     *
+     * @param circuitBreakerState - The bytes32 state of the token of interest.
+     * @param currentWeightComplement - The complement of this token's weight: (_denormWeightSum - token weight)
+     * @return - lower and upper BPT price bounds, which can be directly compared against the current BPT price.
+     */
+    function getCurrentCircuitBreakerBounds(bytes32 circuitBreakerState, uint256 currentWeightComplement)
+        internal
+        pure
+        returns (uint256, uint256)
+    {
+        uint256 referenceBptPrice = circuitBreakerState.decodeUint(
+            _REFERENCE_BPT_PRICE_OFFSET,
+            _REFERENCE_BPT_PRICE_WIDTH
+        );
+        uint256 referenceWeightComplement = circuitBreakerState
+            .decodeUint(_REFERENCE_WEIGHT_COMPLEMENT_OFFSET, _REFERENCE_WEIGHT_COMPLEMENT_WIDTH)
+            .decompress(_REFERENCE_WEIGHT_COMPLEMENT_WIDTH, _MAX_BOUND_PERCENTAGE);
+
+        if (referenceWeightComplement == currentWeightComplement) {
+            // If the weight factor hasn't changed since the circuit breaker was set, we can use the precomputed
+            // boundary expressions.
+            return (
+                referenceBptPrice.mulDown(
+                    circuitBreakerState.decodeUint(_REFERENCE_LOWER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                        _BOUND_RATIO_WIDTH,
+                        _MAX_BOUND_PERCENTAGE
+                    )
+                ),
+                referenceBptPrice.mulUp(
+                    circuitBreakerState.decodeUint(_REFERENCE_UPPER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                        _BOUND_RATIO_WIDTH,
+                        _MAX_BOUND_PERCENTAGE
+                    )
+                )
+            );
+        } else {
+            // Something has changed - either the weight of the token, or the composition of the pool, so we must
+            // retrieve the raw percentage bounds and do the full calculation.
+            (uint256 lowerBoundRatio, uint256 upperBoundRatio) = getBoundaryConversionRatios(
+                circuitBreakerState.decodeUint(_LOWER_BOUND_PCT_OFFSET, _BOUND_PERCENTAGE_WIDTH).decompress(
+                    _BOUND_PERCENTAGE_WIDTH,
+                    _MAX_BOUND_PERCENTAGE
+                ),
+                circuitBreakerState.decodeUint(_UPPER_BOUND_PCT_OFFSET, _BOUND_PERCENTAGE_WIDTH).decompress(
+                    _BOUND_PERCENTAGE_WIDTH,
+                    _MAX_BOUND_PERCENTAGE
+                ),
+                currentWeightComplement
+            );
+
+            // Use these ratios to convert raw percentage bounds to BPT price bounds.
+            // To maximize the valid trading range, round the lower bound down, and the upper bound up.
+            return (referenceBptPrice.mulDown(lowerBoundRatio), referenceBptPrice.mulUp(upperBoundRatio));
```

Comment:
The ratios are not explained on the other return value. I recommend having a common `return` statement outside the `if/else` construct, so that the onl difference is how the ratios are computed.

Also, why do we want to maximize the trading range? That means the breaker could not trip when it should. I'd argue we should be err on the side of being restrictive.